lib/zoho/api/model.rb

```diff
@@ -3,38 +3,81 @@
 module Zoho
   module API
     class Model
-      @@connection ||= nil
-
       class << self
-        def resource_url
-          raise NotImplementedError
-        end
-
-        def connection(_auth: false)
-          @@connection ||=
-            unless @@connection.present?
-              Faraday.new(
-                url: base_url,
-                headers: {
-                  Accept: 'application/json'
-                }
-              ) do |builder|
-                # if _auth
-                #   builder.request :authorization, :basic,
-                #                   Zoho::Credentials.client_id,
-                #                   Zoho::Credentials.client_secret
-                # end
-                # builder.request :json
-                builder.response :json
-                builder.response :logger if Rails.env.development?
-                builder.response :raise_error, include_request: true
-              end
-            end
-        end
-
-        def base_url
+        def connection(auth: false, access_token: nil)
+          url = base_url(auth:)
+          Faraday.new(url:) do |builder|
+            builder.request :authorization, 'Zoho-oauthtoken', access_token if access_token.present?
+            builder.request :json
+            builder.response :json
+            builder.response :logger if Rails.env.development?
+            builder.response :raise_error, include_request: true
+          end
+        end
+
+        def base_url(*)
           'https://www.zohoapis.com'
         end
+
+        def auth_endpoint_url
+          @auth_endpoint_url ||= regional_oauth_endpoint_url
+        end
+
+        def regional_oauth_endpoint_url(region = 'us')
+          unless allowed_regions.include?(region.to_sym)
+            exception = ::LarCity::Errors::Unknown3rdPartyHostError.new('Unsupported 3rd party service region')
+            raise exception
+          end
+
+          response = connection(auth: true).get('/oauth/serverinfo')
+          data = response.body || {}
+          url = data.dig('locations', region.to_s)
+          if valid_http_host?(url)
+            url
```

Comment:
TODO: Don't return the "user input" from here (this is triggering a security issue in code scans)